# It is a regression problem 

## We are doing the machine learning on house price dataset

<table align="center" width=100%>
    <tr>
        </td>
            </td>
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                    <b>Regression modelling on House Price Data
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

# Problem Statement

This dataset is created for the prediction on House Prices from an Indian perspective towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold. 

In [1]:
# import 'Pandas'
# It is utilised to format the data structure. It means it can convert any data structure into pandas data frame.
import pandas as pd

# import 'numpy'
# It is used to find statistical summary.
import numpy as np

# import subpackages of Matplotlib
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

# import 'seaborn'
import seaborn as sns

In [5]:
# Load the data and name it df_house_price
# Load the csv file
# store the data in 'df_houseprice'
df_houseprice = pd.read_csv("C:\\Users\\HP\\Downloads\\House_Price1.csv")

# display first five observation using head()
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

## Data Overview

In [12]:
# Name the Columns 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

 - `price`:Price of the House.
 - `resid_area`:Residential Area where my house is located.

In [15]:
# Basic Info of the Data Set
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**

 - we have 13 numerical column and 3 categorial column
 - we have null values in the data
 - we have total 16 columns in the data

In [18]:
# Shape of the Data
df_houseprice.shape

(506, 16)

**Interpretation**
 - We have total 16 columns.
 - We have 506 observations.

In [23]:
# Summarize the basic statistics of Data
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

**Interpretation**
 - We have outliers in the data, most of the outliers are in age column.

## Data Preprocessing

### a. Data Cleaning

In [32]:
# Show the first five observations
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [34]:
# Rename the Columns
# Step 1 : Show the column names 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [38]:
# Step 2 : Allocate new column names to old column names
# Make the dictionary{'Old column name':'New column name'}
d1={'price':'Price',
    'resid_area':'Residential_Area',
    'air_qual':'Air_Quality',
    'room_num':'Room_Numbers',
    'age':'Age',
    'distance':'Distance',
    'teachers':'Teachers',
    'poor_prop':'Proportion_Of_Poor_Population',
    'airport':'Airport',
    'n_hos_beds':'Number_Of_Hospital_Beds',
    'n_hot_rooms':'Number_Of_Hotel_Rooms',
    'waterbody':'Waterbody',
    'rainfall':'Rainfall',
    'bus_ter':'Bus_Terminal',
    'parks':'Parks',
    'Sold':'Sold'}

In [42]:
df_houseprice = df_houseprice.rename(columns=d1)

In [44]:
# Authenticate the Results
df_houseprice.columns

Index(['Price', 'Residential_Area', 'Air_Quality', 'Room_Numbers', 'Age',
       'Distance', 'Teachers', 'Proportion_Of_Poor_Population', 'Airport',
       'Number_Of_Hospital_Beds', 'Number_Of_Hotel_Rooms', 'Waterbody',
       'Rainfall', 'Bus_Terminal', 'Parks', 'Sold'],
      dtype='object')

### b. Null Value Handling

In [50]:
# Find the valid Null Values
df_houseprice.isnull().sum()

Price                              0
Residential_Area                   0
Air_Quality                        0
Room_Numbers                       0
Age                                0
Distance                           0
Teachers                           0
Proportion_Of_Poor_Population      0
Airport                            0
Number_Of_Hospital_Beds            8
Number_Of_Hotel_Rooms              0
Waterbody                        155
Rainfall                           0
Bus_Terminal                       0
Parks                              0
Sold                               0
dtype: int64

In [54]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                             0.000
Residential_Area                  0.000
Air_Quality                       0.000
Room_Numbers                      0.000
Age                               0.000
Distance                          0.000
Teachers                          0.000
Proportion_Of_Poor_Population     0.000
Airport                           0.000
Number_Of_Hospital_Beds           1.581
Number_Of_Hotel_Rooms             0.000
Waterbody                        30.632
Rainfall                          0.000
Bus_Terminal                      0.000
Parks                             0.000
Sold                              0.000
dtype: float64

In [56]:
# Find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   Residential_Area               506 non-null    float64
 2   Air_Quality                    506 non-null    float64
 3   Room_Numbers                   506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   Proportion_Of_Poor_Population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number_Of_Hospital_Beds        498 non-null    float64
 10  Number_Of_Hotel_Rooms          506 non-null    float64
 11  Waterbody                      351 non-null    object 
 12  Rainfall                       506 non-null    int

In [58]:
# This is a categorical column so we can replace the null with mode.
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [60]:
# Replace the null in that column
df_houseprice.Waterbody = df_houseprice.Waterbody.fillna('River')

In [62]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                            0.000
Residential_Area                 0.000
Air_Quality                      0.000
Room_Numbers                     0.000
Age                              0.000
Distance                         0.000
Teachers                         0.000
Proportion_Of_Poor_Population    0.000
Airport                          0.000
Number_Of_Hospital_Beds          1.581
Number_Of_Hotel_Rooms            0.000
Waterbody                        0.000
Rainfall                         0.000
Bus_Terminal                     0.000
Parks                            0.000
Sold                             0.000
dtype: float64

In [64]:
# In Number_Of_Hospital_Beds contain less than 10% values so we can drop null values
df_houseprice.dropna(inplace=True)

In [66]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                            0.0
Residential_Area                 0.0
Air_Quality                      0.0
Room_Numbers                     0.0
Age                              0.0
Distance                         0.0
Teachers                         0.0
Proportion_Of_Poor_Population    0.0
Airport                          0.0
Number_Of_Hospital_Beds          0.0
Number_Of_Hotel_Rooms            0.0
Waterbody                        0.0
Rainfall                         0.0
Bus_Terminal                     0.0
Parks                            0.0
Sold                             0.0
dtype: float64